### ISSUE TITLE
Deprecate support for typealias qualifiers

### ISSUE THREAD

**Issue by rharter** [evant/kotlin-inject#253]
Using `typealias`es as disambiguating qualifiers is a stretch of the language feature and doesn't align with the `typealias` contract.  A `typealias` allows you to use another name for a thing, whereas a qualifier is meant to disambiguate between two different things represented by the same underlying type.

Additionally, the presence of [inline classes](https://kotlinlang.org/docs/inline-classes.html) eliminates the need for any kind of qualifier, since they provide a toll-free way to disambiguate values based on what they represent, rather than their type.

The suggestion here is to deprecate the use of `typealias`es as qualifiers and instead recommend the use of inline classes (i.e. user defined value classes) for such cases.

**Comment by burningfireplace** [comment]
I see, how you could say that this is a stretch of the original language feature but I don't see how it can cause any harm now or in the future. Inline classes aren't a real `typealias` replacement because you have to "unpack/unwrap" the inline classes to get to the original type. IMHO: inline classes solve another problem that overlaps sometime with this use case.

An optional processor option that throws a warning or error and something like Dagger's qualifier annotations should do the trick for code bases that don't want to use `typealias` to differentiate between different instances of the same type.

**Comment by evant** [comment]
Not a fan of a processor option because turning it on or off could have very subtle effects on how types resolve. And qualifiers are already supported behind the javax.inject option though personally not a fan of their lack of type-safety.

**Comment by JakeWharton** [comment]
> I don't see how it can cause any harm now or in the future.

Using the refactoring option to introduce a typealias for a type which participates in the graph is now a breaking change. This is the opposite of how a typealias is meant to work.

**Comment by burningfireplace** [comment]
Okay, I see now how a typealias could break the DI graph. What about annotated `typealias`es? As far as I understand it after a glimpse at the ksp compiler source and a google search, it should be possible to do something like this:

``` Kotlin
@Qualifier
typealias Dep1 = Dep

@Qualifier
typealias Dep2 = Dep

@Component
abstract class MyComponent {
    @Provides
    fun dep1(): Dep1 = Dep("one")

    @Provides
    fun dep2(): Dep2 = Dep("two")

    @Provides
    fun provides(dep1: Dep1, dep2: Dep2): Thing = Thing(dep1, dep2)
}

@Inject
class InjectedClass(dep1: Dep1, dep2: Dep2)
```

**Comment by damianw** [comment]
IMO: A more natural pattern, which would be consistent with things like kotlinx-serialization, would be to support qualifier annotations directly on types. That would be a bit closer to how Dagger does it too, and allows for more expressive options than solely relying on types or type aliases. You could still use type aliases in that case, the same [as they suggest for kotlinx-serialization](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serializers.md#specifying-serializer-globally-using-typealias), but in a way that doesn't abuse the language feature.

For example, I have seen use cases come up where it is more natural to qualify something by a string name rather than by an actual type (e.g. a service with a particular name); that is possible with Dagger (using `Named`) but not possible here due to lack of support for richer qualifiers (without generating a bespoke typealias or wrapper class as proposed in the OP).

I would propose adding a `Qualifier` annotation as Dagger has, to be applied to annotation types that themselves apply to types:

```kotlin
@Target(AnnotationTarget.ANNOTATION_CLASS)
annotation class Qualifier
```

Such qualifiers could be defined as so:

```kotlin
@Qualifier
@Target(AnnotationTarget.TYPE)
annotation class Named(val value: String)
```

That would then allow for flexibility such as:

```kotlin
data class Dep(val name: String)

typealias Dep1 = @Named("one") Dep

@Component
abstract class MyComponent {

    // Using typealias allows for shorter names like is already possible.
    @Provides
    fun dep1(): Dep1 = Dep("one")

    // Or, the qualifiers can be used inline.
    @Provides
    fun dep2(): @Named("two") Dep = Dep("two")
}
```

**Comment by damianw** [comment]
@evant Thoughts on the above? Would you be open to a PR exploring this?

**Comment by chrisbanes** [comment]
Why do we need typealiases at all? Can we not rely on qualifiers entirely like Dagger?

``` kotlin
@Component
abstract class MyComponent {
    @Provides
    @Named("two")
    fun dep1() = Dep("one")
}
```

**Comment by chrisbanes** [comment]
We ended up enabling the [javax.inject](https://github.com/evant/kotlin-inject#options) support for this, but as we're building for KMP I had to create a 'fake' annotation `javax.inject.Qualifier` class.

``` kotlin
package javax.inject

annotation class Qualifier
```

``` kotlin
package com.myapp.qualifiers

@Qualifier
annotation class IoCoroutineDispatcher
```

Seems to be working nicely.

**Comment by evant** [comment]
Usage note: I've been trying to use value classes in place of typealias on my side project to see how it feels, the downside is that it makes usages sites a bit more verbose and I fell into this pattern:
```kotlin
value class IODispatcher(val value: CoroutineDispatcher)

@Inject class MyRepository(dispatcher: IODispatcher) {
   private val dispatcher = dispatcher.value
}
```
this may lead credence to including qualifiers to cut down on this. This isn't an argument for one thing or another just thought I'd jot it down.

**Comment by vRallev** [comment]
I agree with Chris and wish qualifiers as annotations would be supported similar to Dagger. We often create qualifiers share them between different scopes, such as `@ForScope(AppScope::class)`. We reuse this qualifier across scopes and types, which safes a lot of boilerplate. By forcing typealiases or inline classes I have to create a new qualifier for each scope and each type, which is a lot of boilerplate and it's harder to enforce consistency. The same is true for scopes, e.g. we use `@SingleIn(AppScope::class)`.

**Comment by evant** [comment]
@damianw's solution seems like the way to go. Qualifiers will be directly supported and you can continue to use typealies with the qualifier if you so choose. The steps to migrate to this will be:

1. Add `@Qualifier` to the set of annotations provided, the javax-specific support will be extended to look for those as well, allowing you to use them without a special flag and awkward package name.
2. If conflicting types are found that are only differentiated by a typealias a warning will be emitted asking to add a qualifier.
4. In a future version this will be upgraded to the same error you get today if you provide the same type multiple times.
5. Since the typealis will be treated like the original type, you can mix and match them while injecting. (This won't be enabled until the conflict becomes an error to ensure there's no surprises on what is actually injected)

Note: other annotations like `@Named` won't be added at this time, they are just a one-liner in your project. May be considered in the future.

**Comment by evant** [comment]
Oh something important I just realized. The above assume the qualifier is annotating a type, but javax qualifiers are specified (and currently implemented) to annotate a parameter, method, or field instead. I wonder if this is a side-effect of placing annotations on a type not being a thing before java 8. So we need to figure out where to actually support these:

1. Only on types (may cause confusion as that's different from the existing behavior for qualifiers, also having `me.tatarka.Qualifier` work differently than `javax.inject.Qualifier` is ew, but if the javax one was also changed it would make it less useful for migrating a codebase)
2. Both on types and declarations (may lead to inconsistent placement of annotations)
3. Only on declarations (would give up annotating type aliases)

Finally, annotating types also allows things like:

```kotlin
@Inject MyClass(val list: List<@Named("name") String>)

@Component abstract class MyComponent {
  abstract val myClass: MyClass

  @Provides fun provideNamedList(): List<@Named("name") String> = listOf("one", "two")
} 
```
which is certainly strange, but maybe it's fine?

**Comment by chrisbanes** [comment]
I like option 2. Can always disable one of the option if it becomes a platform a problem in the future

**Comment by evant** [comment]
You also get fun puzzles, like which ones of these resolve to the same type, and which are distinct?

```kotlin
@Inject MyClass(@Named("a") arg: String)
@Inject MyClass(arg: @Named("a") String)
@Inject MyClass(@Named("a") arg: @Named("a") String)
```

**Comment by evant** [comment]
Hm, would limiting each type to one qualifier be reasonable? We could treat the first 2 above the same and error on the third with multiple qualifier declarations. 🤔

**Comment by chrisbanes** [comment]
> Hm, would limiting each type to one qualifier be reasonable?

I'd say yes. You can always create your own composite qualifier if you need (I think)

**Comment by evant** [comment]
As mentioned above, `kotlinx-serialization` may be a good source of [inspiration](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/-serializable/) here as they *also* allow the annotation to be placed on both types and declarations. I can spend some time looking into how they handle edge-cases.

**Comment by evant** [comment]
Answer to the above: in kotlinx the annotation on the declaration takes precedence over the annotation on the type, so if you have both only the declaration one will be respected. It also allows you to annotate types in generics, ex: `List<@Serializable(DateAsLongSerializer::class) Date>` which is interesting. Granted our usecase is bit different but it was an interesting investigation.

Anyway, I think to start we'll support:

```kotlin
@Inject MyClass(@Named("a") arg: String)
```
as equivalent to
```
@Inject MyClass(arg: @Named("a") String)
```
and
```kotlin
@Inject MyClass(@Named("a") arg: @Named("b") String)
```
will throw an error as it's ambiguous (regardless if the qualifier is actually the same or not)

and
```kotlin
@Inject MyClass(args: List<@Named("a") String>)
```
will throw an error as not support. (This may be something to add later but not convinced it's actually all that useful at this time).
